File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/C-Гипотеза_Гольдбаха/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    sieve: list[bool] = [False] * (n + 1)
    prime_numbers: list[int] = []
    for d in range(2, n + 1):
        if not sieve[d]:
            prime_numbers.append(d)
            for next_d in range(d, n + 1, d): sieve[next_d] = True
    seen: set[int] = set()
    x = y = -1
    for prime in prime_numbers:
        seen.add(prime)
        if n - prime in seen:
            x, y = n - prime, prime
    sys.stdout.write('{} {}\n'.format(x, y))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/C-Гипотеза_Гольдбаха/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <cmath>


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> primeNumbers;
    std::vector<bool> sieve(n + 1, false);
    for (int d = 2; d <= n; ++d) {
        if (!sieve[d]) {
            primeNumbers.push_back(d);
            for (int nextD = d; nextD <= n; nextD += d) sieve[nextD] = true;
        }
    }
    std::unordered_set<int> seen;
    int x = -1, y = -1;
    for (int prime : primeNumbers) {
        seen.insert(prime);
        if (seen.count(n - prime)) {
            x = n - prime, y = prime;
        }
    }
    std::printf("%d %d\n", x, y);
}


int main() {
    solution();
}